subsys/bluetooth/host/hci_core.c

```diff
@@ -7185,6 +7185,153 @@ bool bt_addr_le_is_bonded(uint8_t id, const bt_addr_le_t *addr)
+		return -EINVAL;
+	} else if (!atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
+		return -EINVAL;
+	}
+
+	if (param->interval_min < 0x0006 ||
+	    param->interval_max < 0x0006 ||
+	    param->interval_min > param->interval_max) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_PARAM, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = sys_cpu_to_le16(adv->handle);
+	cp->min_interval = sys_cpu_to_le16(param->interval_min);
+	cp->max_interval = sys_cpu_to_le16(param->interval_max);
+
+	if (param->options & BT_LE_PER_ADV_OPT_USE_TX_POWER) {
+		cp->props |= BT_HCI_LE_ADV_PROP_TX_POWER;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_PARAM, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	atomic_set_bit(adv->flags, BT_PER_ADV_PARAMS_SET);
+
+	return 0;
+}
+
+int bt_le_per_adv_set_data(const struct bt_le_ext_adv *adv,
+			   const struct bt_data *ad, size_t ad_len)
+{
+	struct bt_hci_cp_le_set_per_adv_data *cp;
+	struct net_buf *buf;
+	struct bt_ad d = { .data = ad, .len = ad_len };
+	int err;
+
+	if (!atomic_test_bit(adv->flags, BT_PER_ADV_PARAMS_SET)) {
+		return -EINVAL;
+	}
+
+	if (!ad_len || !ad) {
+		return -EINVAL;
+	}
+
+	if (ad_len > BT_HCI_LE_PER_ADV_FRAG_MAX_LEN) {
+		return -EINVAL;
+	}
+
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_DATA, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	/* TODO: If data is longer than what the controller can manage,
+	* split the data. Read size from controller on boot.
+	*/
+	cp->op = BT_HCI_LE_PER_ADV_OP_COMPLETE_DATA;
+
+	err = set_data_add(cp->data, BT_HCI_LE_PER_ADV_FRAG_MAX_LEN, &d, 1,
+			   &cp->len);
+	if (err) {
+		return err;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_DATA, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int bt_le_per_adv_enable(struct bt_le_ext_adv *adv, bool enable)
+{
+	struct bt_hci_cp_le_set_per_adv_enable *cp;
+	struct net_buf *buf;
+	struct cmd_state_set state;
+	int err;
+
+	/* TODO: We could setup some default ext adv params if not already set*/
+	if (!atomic_test_bit(adv->flags, BT_PER_ADV_PARAMS_SET)) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(adv->flags, BT_PER_ADV_ENABLED)) {
+		return -EALREADY;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_ENABLE, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = adv->handle;
+	cp->enable = enable ? 1 : 0;
+
+	cmd_state_set_init(&state, adv->flags, BT_PER_ADV_ENABLED, enable);
+	cmd(buf)->state = &state;
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_ENABLE, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+int bt_le_per_adv_start(struct bt_le_ext_adv *adv)
```

Comment:
Is there any reason this requires new APIs instead of just extending the existing APIs? From API point of view that would have been simpler since these APIs more or less work the same apart from the intervals.

Reply:
They can be started and stopped independently of each other. I think this approach gives a lot of flexibility.

Reply:
I agree, it's really a full other mechanism beyond just regular extended adv, this is more akin to a real connection.

Reply:
I did spend some time considering this as well. My challenge was that I did not understand the use case to be able to advertise these independently. 

The advertiser is not aware of any syncs, so it isn't able to determine if/when it makes sense to advertise both or just the PA, without a user's input. 

The only use case I can think of so far, is where the user (a person) e.g. presses a button to enable extended advertising for some time such that new devices may sync to the PA (somewhat similar to how some BT speakers have the same w.r.t. connections). In such a use case, we need to be able to enable/disable extended and periodic advertising independently. 

In summary: I think this API makes the most sense, as it supports some use cases that otherwise would not be possible.

Reply:
Well I was thinking that the instances would be enabled independently anyway, as I would imagine the start API can check what type of instances there are and enable them accordinly, same thing with stop.

As the syncs, or perhaps the hability of advertising the same data on both PA and EXT format, I guess that would be an argument to have an API which can control them together, meaning one can create instances that advertise on both without having to operate on differente instance and APIs, we would just need a new type for that.

Reply:
They don't have to be two instances though. It can be the same instance, having regular advertising data and periodic advertising data.

Reply:
In the current implementation of this PR, there shall be only one instance if you want to enable PA. You cannot have `struct bt_le_ext_adv` advertising the EA (ext adv) and another `struct bt_le_ext_adv`  advertising the PA. 

We could introduce some API, as @Vudentz suggests, to bind the two instances and/or control both of them in a single API call, but I think that would become fairly complex to handle in the end. 

> I would imagine the start API can check what type of instances there are and enable them accordinly, same thing with stop.
Unless the start/stop functions take a type as an argument, I don't see how this will work.